pkg/algorithm/hived_algorithm.go

```diff
@@ -55,15 +55,39 @@ type HivedAlgorithm struct {
 	vcFreeCellNum map[api.VirtualClusterName]map[CellChain]map[CellLevel]int32
 	// total number of preassigned free cells in all the VCs of each cell type
 	allVCFreeCellNum map[CellChain]map[CellLevel]int32
-	// number of cells left in the physical cluster of each cell type
-	// (i.e., the cell is either in the free list or could be obtained by splitting such a free cell.)
+	// Number of cells left in the physical cluster of each cell type
+	// (the cell is either in the free list or could be obtained by splitting an ancestor that is in the free list).
+	// For example, we have a level-2 cell in the free list (which can be split into 2 level-1 cells).
+	// Then the left cell numbers will be level-2: 1, level-1: 2.
 	totalLeftCellNum map[CellChain]map[CellLevel]int32
 
-	// number of bad free cells in the physical cluster of each cell type
-	badFreeCellNum map[CellChain]map[CellLevel]int32
-	// number of free cells in each VC that are doomed to be bad (i.e., when the healthy
-	// free cells in the physical cluster is fewer that the VC's free cells)
-	vcDoomedBadCellNum map[api.VirtualClusterName]map[CellChain]map[CellLevel]int32
+	// badFreeCells, vcDoomedBadCells, and allVCDoomedBadCellNum are used to track bad cells.
+	// Note that a cell is bad if ANY of its children is bad; so a cell may also contain healthy children.
+	// A preassigned cell in a VC is "doomed to be bad" when the healthy free cells in the physical cluster
+	// is fewer than the VC's free cells (thus certain free cells in the VC will be inevitably bound
+	// to bad physical cells at this moment despite the dynamic cell binding).
+	// Marking doomed bad cells can help the intra-VC scheduler avoid finding placements that will
+	// be inevitably mapped to bad physical nodes. Status of doomed bad cells can also be exposed
+	// to users, so that users can know if their VCs have bad nodes currently.
+
+	// bad free cells in the physical cluster of each cell type (same definition of "free" as above)
+	badFreeCells map[CellChain]ChainCellList
+	// free cells in each VC that are doomed to be bad
+	vcDoomedBadCells map[api.VirtualClusterName]map[CellChain]ChainCellList
+	// number of doomed bad cells in all the VCs of each cell type
+	allVCDoomedBadCellNum map[CellChain]map[CellLevel]int32
+
+	// Whether we should avoid binding cells to nodes that are not within K8s suggested nodes (note that
+	// we always avoid using bad nodes; avoiding non-suggested nodes is optional can be configured).
+	//
+	// Note that we do not mark virtual cells as "doomed to be bound to non-suggested nodes"
+	// (like what we do for "doomed to be bound to bad cells"),
+	// because suggested nodes are not stable: they are provided by K8s for each pod to be scheduled,
+	// and may change across different pods. The consequence is that the intra-VC scheduler may choose
```

Comment:
> The consequence is that  [](start = 41, length = 25)

 The consequence is that, even if ignoreK8sSuggestedNodes is false, the intra-VC scheduler may choose...